packages/plugins/jwt/src/plugin.ts

```diff
@@ -0,0 +1,234 @@
+          throw badRequestError(`Failed to decode authentication token.`);
+        }
+
+        if (!decodedToken) {
+          throw badRequestError(`Failed to decode authentication token, payload is missing.`);
+        }
+
+        // Fetch the signing key based on the key id.
+        const signingKey = await getSigningKey(decodedToken?.header.kid);
+
+        if (!signingKey) {
+          throw badRequestError(`Signing key is not available for this kind of token.`);
+        }
+
+        // Verify the token with the signing key.
+        const verified = await verify(
+          logger,
+          token,
+          signingKey,
+          normalizedOptions.tokenVerification,
+        );
+
+        if (!verified) {
+          throw unauthorizedError(`Unauthenticated`);
+        }
+
+        if (verified) {
+          // Link the verified payload with the request (see `onContextBuilding` for the reading part)
+          payloadByRequest.set(payload.request, verified);
+        }
+      } catch (e) {
+        // User-facing errors should be handled based on the configuration.
+        // These errors are handled based on the value of "reject.invalidToken" config.
+        if (e instanceof GraphQLError) {
+          if (normalizedOptions.reject.invalidToken) {
+            throw e;
+          }
+
+          return;
+        }
+
+        // Server/internal errors should be thrown, so they can be handled by the error handler and be masked.
+        throw e;
+      }
+    },
+    onContextBuilding({ context, extendContext }) {
+      if (context.request == null) {
+        throw new Error(
+          'Request is not available on context! Make sure you use this plugin with GraphQL Yoga.',
+        );
+      }
+
+      // Get the payload and inject it into the GraphQL context.
+      const payload = payloadByRequest.get(context.request);
+      if (payload !== null && normalizedOptions.extendContextFieldName) {
+        extendContext({
+          [normalizedOptions.extendContextFieldName]: payload,
+        });
+      }
+    },
+  };
+}
+
+function badRequestError(message: string, options?: Parameters<typeof createGraphQLError>[1]) {
+  return createGraphQLError(message, {
+    extensions: {
+      http: {
+        status: 400,
+      },
+    },
+    ...options,
+  });
+}
+
+function unauthorizedError(message: string, options?: Parameters<typeof createGraphQLError>[1]) {
+  return createGraphQLError(message, {
+    extensions: {
+      http: {
+        status: 401,
+      },
+    },
+    ...options,
+  });
+}
+
+function verify(
+  logger: YogaLogger,
+  token: string,
+  signingKey: string,
+  options: VerifyOptions | undefined,
+) {
+  return new Promise((resolve, reject) => {
+    jsonwebtoken.verify(token, signingKey, options, (err, result) => {
+      if (err) {
+        logger.warn(`Failed to verify authentication token: `, err);
+        reject(unauthorizedError('Failed to verify authentication token. Verification failed.'));
```

Comment:
Should we give that many details? Verification/Authorization failed should be enough IMHO.

Reply:
Fixed.

I also have a few places where I'm calling `badRequestError` with some details (and error 400). Do you think I should reduce the amount of details there as well?

Reply:
I think we should log the information, but not send it back to the developer.

Reply:
Sure. Makes sense. Fixing